src/main/java/com/google/devtools/build/lib/bazel/repository/cache/RepoContentsCache.java

```diff
@@ -128,4 +165,72 @@ private static String getNextCounterInDir(Path entryDir) throws IOException {
       return counter;
     }
   }
+
+  public void acquireSharedLock() throws IOException {
+    Preconditions.checkState(path != null);
+    Preconditions.checkState(sharedLock == null, "this process already has the shared lock");
+    sharedLock = FileSystemLock.get(path.getRelative(LOCK_PATH), LockMode.SHARED);
+  }
+
+  public void releaseSharedLock() throws IOException {
+    Preconditions.checkState(sharedLock != null);
+    sharedLock.close();
+    sharedLock = null;
+  }
+
+  /**
+   * Creates a garbage collection {@link IdleTask} that deletes cached repos who are last accessed
+   * more than {@code maxAge} ago, with an idle delay of {@code idleDelay}.
+   */
+  public IdleTask createGcIdleTask(Duration maxAge, Duration idleDelay) {
+    Preconditions.checkState(path != null);
+    return new IdleTask() {
+      @Override
+      public String displayName() {
+        return "Repo contents cache garbage collection";
+      }
+
+      @Override
+      public Duration delay() {
+        return idleDelay;
+      }
+
+      @Override
+      public void run() throws InterruptedException, IdleTaskException {
+        try {
+          try (var lock = FileSystemLock.tryGet(path.getRelative(LOCK_PATH), LockMode.EXCLUSIVE)) {
+            // If we can't grab the lock, abort GC. Someone will come along later.
+            runGc(maxAge);
+          }
+        } catch (IOException e) {
+          throw new IdleTaskException(e);
+        }
+      }
+    };
+  }
+
+  private void runGc(Duration maxAge) throws InterruptedException, IOException {
+    Preconditions.checkState(path != null);
+    Instant cutoff = Instant.now().minus(maxAge);
+    for (Path entryDir : path.getDirectoryEntries()) {
+      if (!entryDir.isDirectory()) {
+        continue;
+      }
+      for (Path recordedInputsFile : entryDir.getDirectoryEntries()) {
+        if (!recordedInputsFile.getBaseName().endsWith(RECORDED_INPUTS_SUFFIX)) {
+          continue;
+        }
+        if (Thread.interrupted()) {
+          throw new InterruptedException();
+        }
+
+        if (Instant.ofEpochMilli(recordedInputsFile.getLastModifiedTime()).isBefore(cutoff)) {
+          // Sorry buddy.
+          var candidateRepo = CandidateRepo.fromRecordedInputsFile(recordedInputsFile);
+          candidateRepo.contentsDir.deleteTree();
+          recordedInputsFile.delete();
```

Comment:
```suggestion
          recordedInputsFile.delete();
          candidateRepo.contentsDir.deleteTree();
```
seems safer when interrupted. But it would also require deleting those repo dirs that no longer have a corresponding recorded inputs file.

Reply:
Is it possible to have a well-known "trash directory" that you atomically rename the directories to delete into before deleting them? This has two advantages:

1. You can't leave a partial directory behind in a place where it might be accidentally interpreted as a complete repo
2. Instead of looking for entries that correspond to incomplete deletes, the GC can simply delete the entire trash directory if it exists

Reply:
I added a trash directory (very good idea!). Was a bit tricky to avoid name conflicts but I think I got it done. PTAL!